File: frontend/lib/models/ticket.dart
// lib/models/ticket.dart
class Ticket {
  final String id;
  final String userId; // corresponds to 'user' in backend
  final String issue;
  final String status; // "open", "in_progress", "resolved", "closed"
  final String role; // "patient", "doctor", "admin", "superadmin"
  final DateTime? solvedAt;
  final String review;
  final List<String> visibleTo;
  final bool suspended;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Ticket({
    required this.id,
    required this.userId,
    required this.issue,
    required this.status,
    required this.role,
    this.solvedAt,
    this.review = '',
    this.visibleTo = const ["patient", "doctor", "admin", "superadmin"],
    this.suspended = false,
    this.createdAt,
    this.updatedAt,
  });

  factory Ticket.fromJson(Map<String, dynamic> json) {
    DateTime? parseDateTime(dynamic value) {
      if (value == null) return null;
      
      if (value is Map<String, dynamic>) {
        final seconds = value['_seconds'] as int? ?? 0;
        final nanoseconds = value['_nanoseconds'] as int? ?? 0;
        return DateTime.fromMillisecondsSinceEpoch(
          seconds * 1000 + (nanoseconds ~/ 1000000),
        );
      } else if (value is String) {
        return DateTime.tryParse(value);
      }
      return null;
    }

    return Ticket(
      id: json['id'] ?? json['_id'] ?? '',
      userId: json['user'] ?? '',
      issue: json['issue'] ?? '',
      status: json['status'] ?? 'open',
      role: json['role'] ?? '',
      solvedAt: parseDateTime(json['solvedAt']),
      review: json['review'] ?? '',
      visibleTo: json['visibleTo'] != null
          ? List<String>.from(json['visibleTo'])
          : ["patient", "doctor", "admin", "superadmin"],
      suspended: json['suspended'] ?? false,
      createdAt: parseDateTime(json['createdAt']),
      updatedAt: parseDateTime(json['updatedAt']),
    );
  }
}

File: backend/Models/Ticket.js
const { db } = require("../firebase"); // Import shared Firebase instance

const ticketsCollection = db.collection("tickets");

const ROLES = ["patient", "doctor", "admin", "superadmin"];
const STATUSES = ["open", "in_progress", "resolved", "closed"];

class Ticket {
  constructor({
    user,
    issue,
    status = "open",
    role,
    solvedAt = null,
    review = "",
    visibleTo = ["patient", "doctor", "admin", "superadmin"],
    suspended = false,
  }) {
    if (typeof user !== "string")
      throw new Error("Invalid user: must be a Firestore document reference");
    if (typeof issue !== "string")
      throw new Error("Invalid issue: must be a string");
    if (!STATUSES.includes(status))
      throw new Error(
        `Invalid status: ${status}. Allowed: ${STATUSES.join(", ")}`
      );
    if (!ROLES.includes(role))
      throw new Error(`Invalid role: ${role}. Allowed: ${ROLES.join(", ")}`);
    if (solvedAt !== null && !(solvedAt instanceof Date))
      throw new Error("Invalid solvedAt: must be a Date object or null");
    if (typeof review !== "string")
      throw new Error("Invalid review: must be a string");
    if (!Array.isArray(visibleTo))
      throw new Error("Invalid visibleTo: must be an array of strings");
    if (typeof suspended !== "boolean")
      throw new Error("Invalid suspended: must be a boolean");

    this.user = user;
    this.issue = issue;
    this.status = status;
    this.role = role;
    this.solvedAt = solvedAt;
    this.review = review;
    this.visibleTo = visibleTo;
    this.suspended = suspended;
    this.createdAt = new Date();
    this.updatedAt = new Date();
  }

  async save() {
    try {
      const docRef = await ticketsCollection.add({ ...this });
      return docRef.id;
    } catch (error) {
      throw new Error("Error saving ticket: " + error.message);
    }
  }
}

module.exports = Ticket;
